charon/src/ast/types/vars.rs

```diff
@@ -0,0 +1,172 @@
+//! Type-level variables. There are 4 kinds of variables at the type-level: regions, types, const
+//! generics and trait clauses. The relevant definitions are in this module.
+use derive_visitor::{Drive, DriveMut};
+use serde::{Deserialize, Serialize};
+
+use crate::ast::*;
+
+/// The index of a binder, counting from the innermost. See [`DeBruijnVar`] for details.
+#[derive(
+    Debug,
+    PartialEq,
+    Eq,
+    Copy,
+    Clone,
+    Hash,
+    PartialOrd,
+    Ord,
+    Serialize,
+    Deserialize,
+    Drive,
+    DriveMut,
+)]
+#[serde(transparent)]
+pub struct DeBruijnId {
+    pub index: usize,
+}
+
+/// Bound region variable.
+///
+/// **Important**:
+/// ==============
+/// Similarly to what the Rust compiler does, we use De Bruijn indices to
+/// identify *groups* of bound variables, and variable identifiers to
+/// identity the variables inside the groups.
+///
+/// For instance, we have the following:
+/// ```text
+///                     we compute the De Bruijn indices from here
+///                            VVVVVVVVVVVVVVVVVVVVVVV
+/// fn f<'a, 'b>(x: for<'c> fn(&'a u8, &'b u16, &'c u32) -> u64) {}
+///      ^^^^^^         ^^       ^       ^        ^
+///        |      De Bruijn: 0   |       |        |
+///  De Bruijn: 1                |       |        |
+///                        De Bruijn: 1  |    De Bruijn: 0
+///                           Var id: 0  |       Var id: 0
+///                                      |
+///                                De Bruijn: 1
+///                                   Var id: 1
+/// ```
+#[derive(Debug, PartialEq, Eq, Copy, Clone, Hash, PartialOrd, Ord, Serialize, Deserialize)]
+pub enum DeBruijnVar<Bound, Free> {
+    /// A variable attached to the nth binder, counting from the inside.
+    Bound(DeBruijnId, Bound),
+    /// A variable attached to an implicit binder outside all other binders. This is not present in
+    /// translated code, and only provided as a convenience for convenient variable manipulation.
+    Free(Free),
+}
```

Comment:
I just realized I'm not super happy with the name `DeBruijnVar`: to me a de Bruijn var is necessarily a bound var, while here it actually refers to a bound var or a free var. I'm aware the name `Var` is already taken. Maybe we should use `VarDecl` (for the current `Var` type) and `Var` (for the current `DeBruijnVar`)?

Reply:
(and yes, it does make sense to see a free var as bound outside all binders: it's just not the standard interpretation I believe)

Reply:
I was thiking the same. The current `Var` should be called `Local` anyway, I thought I had renamed it already.

Reply:
I'll add it to the [list of renames](https://github.com/AeneasVerif/charon/issues/287)